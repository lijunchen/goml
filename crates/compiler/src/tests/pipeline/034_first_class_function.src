fn double(x: int) -> int {
    x * 2
}

fn increment(x: int) -> int {
    x + 1
}

fn apply_once(f: (int) -> int, value: int) -> int {
    f(value)
}

fn compose(f: (int) -> int, g: (int) -> int, value: int) -> int {
    f(g(value))
}

fn main() {
    let local = double in
    let first = apply_once(local, 4) in
    let composed = compose(double, increment, first) in
    let closure_apply = |value: int| apply_once(increment, value) in
    let closure_result = closure_apply(composed) in
    let global_invoker = |func_to_call: (int) -> int, value: int| apply_once(func_to_call, value) in
    let invoked_with_global = global_invoker(double, 3) in
    let composer_closure = |value: int| compose(double, increment, value) in
    let composed_by_closure = composer_closure(5) in
    let printer = string_println in
    let _ = printer(int_to_string(composed)) in
    let _ = printer(int_to_string(closure_result)) in
    let _ = printer(int_to_string(invoked_with_global)) in
    let _ = printer(int_to_string(composed_by_closure)) in
    ()
}
