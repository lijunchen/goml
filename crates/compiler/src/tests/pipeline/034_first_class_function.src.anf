fn double(x/0: int) -> int {
  int_mul(x/0, 2)
}

fn increment(x/1: int) -> int {
  int_add(x/1, 1)
}

fn apply_once(f/2: (int) -> int, value/3: int) -> int {
  f/2(value/3)
}

fn compose(f/4: (int) -> int, g/5: (int) -> int, value/6: int) -> int {
  let t10 = g/5(value/6) in
  f/4(t10)
}

fn main() -> unit {
  let local/7 = double in
  let first/8 = apply_once(local/7, 4) in
  let composed/9 = compose(double, increment, first/8) in
  let closure_apply/11 = ClosureEnv0 {} in
  let closure_result/12 = __closure_apply4(closure_apply/11, composed/9) in
  let global_invoker/15 = ClosureEnv1 {} in
  let invoked_with_global/16 = __closure_apply6(global_invoker/15, double, 3) in
  let composer_closure/18 = ClosureEnv2 {} in
  let composed_by_closure/19 = __closure_apply8(composer_closure/18, 5) in
  let printer/20 = string_println in
  let t11 = int_to_string(composed/9) in
  let mtmp0 = printer/20(t11) in
  let t12 = int_to_string(closure_result/12) in
  let mtmp1 = printer/20(t12) in
  let t13 = int_to_string(invoked_with_global/16) in
  let mtmp2 = printer/20(t13) in
  let t14 = int_to_string(composed_by_closure/19) in
  let mtmp3 = printer/20(t14) in
  ()
}

fn __closure_apply4(env5: ClosureEnv0, value/10: int) -> int {
  apply_once(increment, value/10)
}

fn __closure_apply6(env7: ClosureEnv1, func_to_call/13: (int) -> int, value/14: int) -> int {
  apply_once(func_to_call/13, value/14)
}

fn __closure_apply8(env9: ClosureEnv2, value/17: int) -> int {
  compose(double, increment, value/17)
}