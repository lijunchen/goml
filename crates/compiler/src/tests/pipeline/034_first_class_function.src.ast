fn double(x: int) -> int {
    x * 2
}

fn increment(x: int) -> int {
    x + 1
}

fn apply_once(f: (int) -> int, value: int) -> int {
    f(value)
}

fn compose(f: (int) -> int, g: (int) -> int, value: int) -> int {
    f(g(value))
}

fn main() {
    let local = double in
    let first = apply_once(local, 4) in
    let composed = compose(double, increment, first) in
    let printer = string_println in
    let _ = printer(int_to_string(composed)) in
    ()
}

