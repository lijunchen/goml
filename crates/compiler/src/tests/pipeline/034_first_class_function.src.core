fn double(x/0: int) -> int {
  int_mul(x/0, 2)
}

fn increment(x/1: int) -> int {
  int_add(x/1, 1)
}

fn apply_once(f/2: (int) -> int, value/3: int) -> int {
  f/2(value/3)
}

fn compose(f/4: (int) -> int, g/5: (int) -> int, value/6: int) -> int {
  f/4(g/5(value/6))
}

fn main() -> unit {
  let local/7 = double in
  let first/8 = apply_once(local/7, 4) in
  let composed/9 = compose(double, increment, first/8) in
  let closure_apply/11 = |value/10: int| => apply_once(increment, value/10) in
  let closure_result/12 = closure_apply/11(composed/9) in
  let global_invoker/15 = |func_to_call/13: (int) -> int, value/14: int| => apply_once(func_to_call/13, value/14) in
  let invoked_with_global/16 = global_invoker/15(double, 3) in
  let composer_closure/18 = |value/17: int| => compose(double, increment, value/17) in
  let composed_by_closure/19 = composer_closure/18(5) in
  let printer/20 = string_println in
  let mtmp0 = printer/20(int_to_string(composed/9)) in
  let mtmp1 = printer/20(int_to_string(closure_result/12)) in
  let mtmp2 = printer/20(int_to_string(invoked_with_global/16)) in
  let mtmp3 = printer/20(int_to_string(composed_by_closure/19)) in
  ()
}