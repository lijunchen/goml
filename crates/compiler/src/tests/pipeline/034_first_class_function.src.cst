FILE@0..463
  FN@0..40
    FnKeyword@0..2 "fn"
    Whitespace@2..3 " "
    Lident@3..9 "double"
    PARAM_LIST@9..18
      LParen@9..10 "("
      PARAM@10..16
        Lident@10..11 "x"
        Colon@11..12 ":"
        Whitespace@12..13 " "
        TYPE_INT@13..16
          IntKeyword@13..16 "int"
      RParen@16..17 ")"
      Whitespace@17..18 " "
    Arrow@18..20 "->"
    Whitespace@20..21 " "
    TYPE_INT@21..25
      IntKeyword@21..24 "int"
      Whitespace@24..25 " "
    BLOCK@25..40
      LBrace@25..26 "{"
      Whitespace@26..31 "\n    "
      EXPR_BINARY@31..37
        EXPR_LIDENT@31..33
          Lident@31..32 "x"
          Whitespace@32..33 " "
        Star@33..34 "*"
        Whitespace@34..35 " "
        EXPR_INT@35..37
          Int@35..36 "2"
          Whitespace@36..37 "\n"
      RBrace@37..38 "}"
      Whitespace@38..40 "\n\n"
  FN@40..83
    FnKeyword@40..42 "fn"
    Whitespace@42..43 " "
    Lident@43..52 "increment"
    PARAM_LIST@52..61
      LParen@52..53 "("
      PARAM@53..59
        Lident@53..54 "x"
        Colon@54..55 ":"
        Whitespace@55..56 " "
        TYPE_INT@56..59
          IntKeyword@56..59 "int"
      RParen@59..60 ")"
      Whitespace@60..61 " "
    Arrow@61..63 "->"
    Whitespace@63..64 " "
    TYPE_INT@64..68
      IntKeyword@64..67 "int"
      Whitespace@67..68 " "
    BLOCK@68..83
      LBrace@68..69 "{"
      Whitespace@69..74 "\n    "
      EXPR_BINARY@74..80
        EXPR_LIDENT@74..76
          Lident@74..75 "x"
          Whitespace@75..76 " "
        Plus@76..77 "+"
        Whitespace@77..78 " "
        EXPR_INT@78..80
          Int@78..79 "1"
          Whitespace@79..80 "\n"
      RBrace@80..81 "}"
      Whitespace@81..83 "\n\n"
  FN@83..151
    FnKeyword@83..85 "fn"
    Whitespace@85..86 " "
    Lident@86..96 "apply_once"
    PARAM_LIST@96..126
      LParen@96..97 "("
      PARAM@97..114
        Lident@97..98 "f"
        Colon@98..99 ":"
        Whitespace@99..100 " "
        TYPE_FUNC@100..112
          TYPE_TUPLE@100..106
            TYPE_LIST@100..106
              LParen@100..101 "("
              TYPE_INT@101..104
                IntKeyword@101..104 "int"
              RParen@104..105 ")"
              Whitespace@105..106 " "
          Arrow@106..108 "->"
          Whitespace@108..109 " "
          TYPE_INT@109..112
            IntKeyword@109..112 "int"
        Comma@112..113 ","
        Whitespace@113..114 " "
      PARAM@114..124
        Lident@114..119 "value"
        Colon@119..120 ":"
        Whitespace@120..121 " "
        TYPE_INT@121..124
          IntKeyword@121..124 "int"
      RParen@124..125 ")"
      Whitespace@125..126 " "
    Arrow@126..128 "->"
    Whitespace@128..129 " "
    TYPE_INT@129..133
      IntKeyword@129..132 "int"
      Whitespace@132..133 " "
    BLOCK@133..151
      LBrace@133..134 "{"
      Whitespace@134..139 "\n    "
      EXPR_CALL@139..148
        EXPR_LIDENT@139..140
          Lident@139..140 "f"
        ARG_LIST@140..148
          LParen@140..141 "("
          ARG@141..146
            EXPR_LIDENT@141..146
              Lident@141..146 "value"
          RParen@146..147 ")"
          Whitespace@147..148 "\n"
      RBrace@148..149 "}"
      Whitespace@149..151 "\n\n"
  FN@151..236
    FnKeyword@151..153 "fn"
    Whitespace@153..154 " "
    Lident@154..161 "compose"
    PARAM_LIST@161..208
      LParen@161..162 "("
      PARAM@162..179
        Lident@162..163 "f"
        Colon@163..164 ":"
        Whitespace@164..165 " "
        TYPE_FUNC@165..177
          TYPE_TUPLE@165..171
            TYPE_LIST@165..171
              LParen@165..166 "("
              TYPE_INT@166..169
                IntKeyword@166..169 "int"
              RParen@169..170 ")"
              Whitespace@170..171 " "
          Arrow@171..173 "->"
          Whitespace@173..174 " "
          TYPE_INT@174..177
            IntKeyword@174..177 "int"
        Comma@177..178 ","
        Whitespace@178..179 " "
      PARAM@179..196
        Lident@179..180 "g"
        Colon@180..181 ":"
        Whitespace@181..182 " "
        TYPE_FUNC@182..194
          TYPE_TUPLE@182..188
            TYPE_LIST@182..188
              LParen@182..183 "("
              TYPE_INT@183..186
                IntKeyword@183..186 "int"
              RParen@186..187 ")"
              Whitespace@187..188 " "
          Arrow@188..190 "->"
          Whitespace@190..191 " "
          TYPE_INT@191..194
            IntKeyword@191..194 "int"
        Comma@194..195 ","
        Whitespace@195..196 " "
      PARAM@196..206
        Lident@196..201 "value"
        Colon@201..202 ":"
        Whitespace@202..203 " "
        TYPE_INT@203..206
          IntKeyword@203..206 "int"
      RParen@206..207 ")"
      Whitespace@207..208 " "
    Arrow@208..210 "->"
    Whitespace@210..211 " "
    TYPE_INT@211..215
      IntKeyword@211..214 "int"
      Whitespace@214..215 " "
    BLOCK@215..236
      LBrace@215..216 "{"
      Whitespace@216..221 "\n    "
      EXPR_CALL@221..233
        EXPR_LIDENT@221..222
          Lident@221..222 "f"
        ARG_LIST@222..233
          LParen@222..223 "("
          ARG@223..231
            EXPR_CALL@223..231
              EXPR_LIDENT@223..224
                Lident@223..224 "g"
              ARG_LIST@224..231
                LParen@224..225 "("
                ARG@225..230
                  EXPR_LIDENT@225..230
                    Lident@225..230 "value"
                RParen@230..231 ")"
          RParen@231..232 ")"
          Whitespace@232..233 "\n"
      RBrace@233..234 "}"
      Whitespace@234..236 "\n\n"
  FN@236..463
    FnKeyword@236..238 "fn"
    Whitespace@238..239 " "
    Lident@239..243 "main"
    PARAM_LIST@243..246
      LParen@243..244 "("
      RParen@244..245 ")"
      Whitespace@245..246 " "
    BLOCK@246..463
      LBrace@246..247 "{"
      Whitespace@247..252 "\n    "
      EXPR_LET@252..461
        LetKeyword@252..255 "let"
        Whitespace@255..256 " "
        PATTERN_VARIABLE@256..262
          Lident@256..261 "local"
          Whitespace@261..262 " "
        Eq@262..263 "="
        Whitespace@263..264 " "
        EXPR_LET_VALUE@264..271
          EXPR_LIDENT@264..271
            Lident@264..270 "double"
            Whitespace@270..271 " "
        InKeyword@271..273 "in"
        Whitespace@273..278 "\n    "
        EXPR_LET_BODY@278..461
          EXPR_LET@278..461
            LetKeyword@278..281 "let"
            Whitespace@281..282 " "
            PATTERN_VARIABLE@282..288
              Lident@282..287 "first"
              Whitespace@287..288 " "
            Eq@288..289 "="
            Whitespace@289..290 " "
            EXPR_LET_VALUE@290..311
              EXPR_CALL@290..311
                EXPR_LIDENT@290..300
                  Lident@290..300 "apply_once"
                ARG_LIST@300..311
                  LParen@300..301 "("
                  ARG@301..308
                    EXPR_LIDENT@301..306
                      Lident@301..306 "local"
                    Comma@306..307 ","
                    Whitespace@307..308 " "
                  ARG@308..309
                    EXPR_INT@308..309
                      Int@308..309 "4"
                  RParen@309..310 ")"
                  Whitespace@310..311 " "
            InKeyword@311..313 "in"
            Whitespace@313..318 "\n    "
            EXPR_LET_BODY@318..461
              EXPR_LET@318..461
                LetKeyword@318..321 "let"
                Whitespace@321..322 " "
                PATTERN_VARIABLE@322..331
                  Lident@322..330 "composed"
                  Whitespace@330..331 " "
                Eq@331..332 "="
                Whitespace@332..333 " "
                EXPR_LET_VALUE@333..367
                  EXPR_CALL@333..367
                    EXPR_LIDENT@333..340
                      Lident@333..340 "compose"
                    ARG_LIST@340..367
                      LParen@340..341 "("
                      ARG@341..349
                        EXPR_LIDENT@341..347
                          Lident@341..347 "double"
                        Comma@347..348 ","
                        Whitespace@348..349 " "
                      ARG@349..360
                        EXPR_LIDENT@349..358
                          Lident@349..358 "increment"
                        Comma@358..359 ","
                        Whitespace@359..360 " "
                      ARG@360..365
                        EXPR_LIDENT@360..365
                          Lident@360..365 "first"
                      RParen@365..366 ")"
                      Whitespace@366..367 " "
                InKeyword@367..369 "in"
                Whitespace@369..374 "\n    "
                EXPR_LET_BODY@374..461
                  EXPR_LET@374..461
                    LetKeyword@374..377 "let"
                    Whitespace@377..378 " "
                    PATTERN_VARIABLE@378..386
                      Lident@378..385 "printer"
                      Whitespace@385..386 " "
                    Eq@386..387 "="
                    Whitespace@387..388 " "
                    EXPR_LET_VALUE@388..403
                      EXPR_LIDENT@388..403
                        Lident@388..402 "string_println"
                        Whitespace@402..403 " "
                    InKeyword@403..405 "in"
                    Whitespace@405..410 "\n    "
                    EXPR_LET_BODY@410..461
                      EXPR_LET@410..461
                        LetKeyword@410..413 "let"
                        Whitespace@413..414 " "
                        PATTERN_WILDCARD@414..416
                          WildcardKeyword@414..415 "_"
                          Whitespace@415..416 " "
                        Eq@416..417 "="
                        Whitespace@417..418 " "
                        EXPR_LET_VALUE@418..451
                          EXPR_CALL@418..451
                            EXPR_LIDENT@418..425
                              Lident@418..425 "printer"
                            ARG_LIST@425..451
                              LParen@425..426 "("
                              ARG@426..449
                                EXPR_CALL@426..449
                                  EXPR_LIDENT@426..439
                                    Lident@426..439 "int_to_string"
                                  ARG_LIST@439..449
                                    LParen@439..440 "("
                                    ARG@440..448
                                      EXPR_LIDENT@440..448
                                        Lident@440..448 "composed"
                                    RParen@448..449 ")"
                              RParen@449..450 ")"
                              Whitespace@450..451 " "
                        InKeyword@451..453 "in"
                        Whitespace@453..458 "\n    "
                        EXPR_LET_BODY@458..461
                          EXPR_UNIT@458..461
                            LParen@458..459 "("
                            RParen@459..460 ")"
                            Whitespace@460..461 "\n"
      RBrace@461..462 "}"
      Whitespace@462..463 "\n"