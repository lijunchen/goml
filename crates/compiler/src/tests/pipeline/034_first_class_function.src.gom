package main

import (
    "fmt"
)

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func int_add(x int, y int) int {
    return x + y
}

func int_mul(x int, y int) int {
    return x * y
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

type ClosureEnv0 struct {}

type ClosureEnv1 struct {}

type ClosureEnv2 struct {}

func double(x__0 int) int {
    var ret15 int
    ret15 = int_mul(x__0, 2)
    return ret15
}

func increment(x__1 int) int {
    var ret16 int
    ret16 = int_add(x__1, 1)
    return ret16
}

func apply_once(f__2 func(int) int, value__3 int) int {
    var ret17 int
    ret17 = f__2(value__3)
    return ret17
}

func compose(f__4 func(int) int, g__5 func(int) int, value__6 int) int {
    var ret18 int
    var t10 int = g__5(value__6)
    ret18 = f__4(t10)
    return ret18
}

func main0() struct{} {
    var ret19 struct{}
    var local__7 func(int) int = double
    var first__8 int = apply_once(local__7, 4)
    var composed__9 int = compose(double, increment, first__8)
    var closure_apply__11 ClosureEnv0 = ClosureEnv0{}
    var closure_result__12 int = __closure_apply4(closure_apply__11, composed__9)
    var global_invoker__15 ClosureEnv1 = ClosureEnv1{}
    var invoked_with_global__16 int = __closure_apply6(global_invoker__15, double, 3)
    var composer_closure__18 ClosureEnv2 = ClosureEnv2{}
    var composed_by_closure__19 int = __closure_apply8(composer_closure__18, 5)
    var printer__20 func(string) struct{} = string_println
    var t11 string = int_to_string(composed__9)
    printer__20(t11)
    var t12 string = int_to_string(closure_result__12)
    printer__20(t12)
    var t13 string = int_to_string(invoked_with_global__16)
    printer__20(t13)
    var t14 string = int_to_string(composed_by_closure__19)
    printer__20(t14)
    ret19 = struct{}{}
    return ret19
}

func __closure_apply4(env5 ClosureEnv0, value__10 int) int {
    var ret20 int
    ret20 = apply_once(increment, value__10)
    return ret20
}

func __closure_apply6(env7 ClosureEnv1, func_to_call__13 func(int) int, value__14 int) int {
    var ret21 int
    ret21 = apply_once(func_to_call__13, value__14)
    return ret21
}

func __closure_apply8(env9 ClosureEnv2, value__17 int) int {
    var ret22 int
    ret22 = compose(double, increment, value__17)
    return ret22
}

func main() {
    main0()
}
