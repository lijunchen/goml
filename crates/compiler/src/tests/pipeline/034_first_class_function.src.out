18
19
6
12
